File: packages/design-agent/src/brief.ts
import { mkdirSync, readFileSync, writeFileSync } from 'node:fs'
import path from 'node:path'
import { record, string, stringsAllowEmpty } from './parse.js'

export interface ExplicitBriefAnswer {
  question: string
  answer: string
}

export interface DesignBrief {
  originalRequest: string
  subject: string
  pageType: string
  scope: string
  primaryGoal: string
  audience: string
  offer: string
  primaryAction: string
  requiredContent: string[]
  constraints: string[]
  brandInputs: string[]
  creativeControl: string
  explicitAnswers: ExplicitBriefAnswer[]
  assumptions: string[]
  unresolved: string[]
}

const STRING_FIELDS = [
  'originalRequest',
  'subject',
  'pageType',
  'scope',
  'primaryGoal',
  'audience',
  'offer',
  'primaryAction',
  'creativeControl',
] as const

const STRING_ARRAY_FIELDS = [
  'requiredContent',
  'constraints',
  'brandInputs',
  'assumptions',
  'unresolved',
] as const

function parseDesignBrief(value: unknown): DesignBrief {
  const brief = record(value, 'design brief')
  for (const field of STRING_FIELDS) {
    string(brief[field], `design brief field ${field}`)
  }
  for (const field of STRING_ARRAY_FIELDS) {
    stringsAllowEmpty(brief[field], `design brief field ${field}`)
  }
  if (
    !Array.isArray(brief.explicitAnswers) ||
    !brief.explicitAnswers.every(isExplicitBriefAnswer)
  ) {
    throw new Error('design brief field explicitAnswers must contain question and answer strings')
  }
  // Rebuilt field-by-field like every other parser in this package: the raw
  // cast kept arbitrary model-authored extra keys, which were persisted and
  // re-serialized verbatim into the Build agent's instruction block.
  return {
    originalRequest: string(brief.originalRequest, 'design brief field originalRequest'),
    subject: string(brief.subject, 'design brief field subject'),
    pageType: string(brief.pageType, 'design brief field pageType'),
    scope: string(brief.scope, 'design brief field scope'),
    primaryGoal: string(brief.primaryGoal, 'design brief field primaryGoal'),
    audience: string(brief.audience, 'design brief field audience'),
    offer: string(brief.offer, 'design brief field offer'),
    primaryAction: string(brief.primaryAction, 'design brief field primaryAction'),
    creativeControl: string(brief.creativeControl, 'design brief field creativeControl'),
    requiredContent: stringsAllowEmpty(brief.requiredContent, 'design brief field requiredContent'),
    constraints: stringsAllowEmpty(brief.constraints, 'design brief field constraints'),
    brandInputs: stringsAllowEmpty(brief.brandInputs, 'design brief field brandInputs'),
    assumptions: stringsAllowEmpty(brief.assumptions, 'design brief field assumptions'),
    unresolved: stringsAllowEmpty(brief.unresolved, 'design brief field unresolved'),
    explicitAnswers: brief.explicitAnswers.map((item) => ({
      question: item.question,
      answer: item.answer,
    })),
  }
}

function isExplicitBriefAnswer(value: unknown): value is ExplicitBriefAnswer {
  return (
    typeof value === 'object' &&
    value !== null &&
    'question' in value &&
    typeof value.question === 'string' &&
    'answer' in value &&
    typeof value.answer === 'string'
  )
}

function briefPath(workspacePath: string): string {
  return path.join(workspacePath, '.taste', 'brief.json')
}

export function readDesignBrief(workspacePath: string): DesignBrief {
  return parseDesignBrief(JSON.parse(readFileSync(briefPath(workspacePath), 'utf8')))
}

export function writeDesignBrief(workspacePath: string, value: unknown): DesignBrief {
  const brief = parseDesignBrief(value)
  const outputPath = briefPath(workspacePath)
  mkdirSync(path.dirname(outputPath), { recursive: true })
  writeFileSync(outputPath, `${JSON.stringify(brief, null, 2)}\n`, 'utf8')
  return brief
}

File: crates/harness-design-agent/src/brief.rs
use crate::common::{any_string, object, read_json, string, write_json};
use crate::{DesignError, Result};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::path::{Path, PathBuf};

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExplicitBriefAnswer {
    pub question: String,
    pub answer: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DesignBrief {
    pub original_request: String,
    pub subject: String,
    pub page_type: String,
    pub scope: String,
    pub primary_goal: String,
    pub audience: String,
    pub offer: String,
    pub primary_action: String,
    pub required_content: Vec<String>,
    pub constraints: Vec<String>,
    pub brand_inputs: Vec<String>,
    pub creative_control: String,
    pub explicit_answers: Vec<ExplicitBriefAnswer>,
    pub assumptions: Vec<String>,
    pub unresolved: Vec<String>,
}

pub fn parse_design_brief(value: &Value) -> Result<DesignBrief> {
    let brief = object(value, "design brief")?;
    Ok(DesignBrief {
        original_request: string(
            brief.get("originalRequest"),
            "design brief field originalRequest",
        )?,
        subject: string(brief.get("subject"), "design brief field subject")?,
        page_type: string(brief.get("pageType"), "design brief field pageType")?,
        scope: string(brief.get("scope"), "design brief field scope")?,
        primary_goal: string(brief.get("primaryGoal"), "design brief field primaryGoal")?,
        audience: string(brief.get("audience"), "design brief field audience")?,
        offer: string(brief.get("offer"), "design brief field offer")?,
        primary_action: string(
            brief.get("primaryAction"),
            "design brief field primaryAction",
        )?,
        required_content: brief_strings(brief.get("requiredContent"), "requiredContent")?,
        constraints: brief_strings(brief.get("constraints"), "constraints")?,
        brand_inputs: brief_strings(brief.get("brandInputs"), "brandInputs")?,
        creative_control: string(
            brief.get("creativeControl"),
            "design brief field creativeControl",
        )?,
        explicit_answers: explicit_answers(brief.get("explicitAnswers"))?,
        assumptions: brief_strings(brief.get("assumptions"), "assumptions")?,
        unresolved: brief_strings(brief.get("unresolved"), "unresolved")?,
    })
}

pub fn read_design_brief(workspace_path: &Path) -> Result<DesignBrief> {
    parse_design_brief(&read_json(&brief_path(workspace_path))?)
}

pub fn write_design_brief(workspace_path: &Path, value: &Value) -> Result<DesignBrief> {
    let brief = parse_design_brief(value)?;
    write_json(&brief_path(workspace_path), &brief)?;
    Ok(brief)
}

fn explicit_answers(value: Option<&Value>) -> Result<Vec<ExplicitBriefAnswer>> {
    let values = value.and_then(Value::as_array).ok_or_else(|| {
        DesignError::new(
            "design brief field explicitAnswers must contain question and answer strings",
        )
    })?;
    values
        .iter()
        .map(|value| {
            let answer = value.as_object().ok_or_else(|| {
                DesignError::new(
                    "design brief field explicitAnswers must contain question and answer strings",
                )
            })?;
            Ok(ExplicitBriefAnswer {
                question: any_string(answer.get("question"), "explicit answer question").map_err(
                    |_| {
                        DesignError::new(
                            "design brief field explicitAnswers must contain question and answer strings",
                        )
                    },
                )?,
                answer: any_string(answer.get("answer"), "explicit answer").map_err(|_| {
                    DesignError::new(
                        "design brief field explicitAnswers must contain question and answer strings",
                    )
                })?,
            })
        })
        .collect()
}

fn brief_strings(value: Option<&Value>, field: &str) -> Result<Vec<String>> {
    let values = value.and_then(Value::as_array).ok_or_else(|| {
        DesignError::new(format!("design brief field {field} must be a string array"))
    })?;
    values
        .iter()
        .map(|value| {
            value.as_str().map(str::to_owned).ok_or_else(|| {
                DesignError::new(format!("design brief field {field} must be a string array"))
            })
        })
        .collect()
}

fn brief_path(workspace_path: &Path) -> PathBuf {
    workspace_path.join(".taste").join("brief.json")
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn brief() -> Value {
        json!({
            "originalRequest": "Create a studio site.",
            "subject": "Independent creative studio",
            "pageType": "Marketing site",
            "scope": "Single responsive page",
            "primaryGoal": "Generate qualified enquiries",
            "audience": "Teams seeking creative direction",
            "offer": "Brand and digital design services",
            "primaryAction": "Start a project",
            "requiredContent": ["Selected work"],
            "constraints": [],
            "brandInputs": ["Use the supplied wordmark"],
            "creativeControl": "Agent-led",
            "explicitAnswers": [{"question": "Primary action?", "answer": "Start a project"}],
            "assumptions": [],
            "unresolved": [],
            "ignoredModelKey": "not persisted"
        })
    }

    #[test]
    fn round_trips_only_validated_fields() {
        let workspace = tempfile::tempdir().unwrap();
        let written = write_design_brief(workspace.path(), &brief()).unwrap();
        assert_eq!(read_design_brief(workspace.path()).unwrap(), written);
        let stored = std::fs::read_to_string(brief_path(workspace.path())).unwrap();
        assert!(!stored.contains("ignoredModelKey"));
    }

    #[test]
    fn rejects_incomplete_handoff() {
        let error = parse_design_brief(&json!({"subject": "Studio"})).unwrap_err();
        assert_eq!(
            error.to_string(),
            "design brief field originalRequest must be a non-empty string"
        );
    }
}
